include/aws/mqtt/private/v5/mqtt5_to_mqtt3_adapter_impl.h

```diff
@@ -257,6 +257,12 @@ struct aws_mqtt_client_connection_5_impl {
      */
     struct aws_ref_count internal_refs;
 
+    /*
+     * Set to true if the adapter is in destroying process (internal_ref is 0). The adapter would prevent any
+     * operations if `is_destroying` set to true
+     */
+    bool is_destroying;
```

Comment:
Judgement call: Instead of creating a new member here, we can use `adapter_state` to track the destroy processing. When we start the destroy process (internal ref -> 0), the adapter would not receive any lifecycle events anymore.